be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala

```diff
@@ -0,0 +1,83 @@
+package ch.epfl.pop.decentralized
+
+import akka.NotUsed
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, Timers}
+import akka.event.LoggingReceive
+import akka.stream.scaladsl.Sink
+import ch.epfl.pop.decentralized.Monitor.TriggerHeartbeat
+import ch.epfl.pop.model.network.JsonRpcRequest
+import ch.epfl.pop.model.network.method.ParamsWithMap
+import ch.epfl.pop.pubsub.graph.GraphMessage
+
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+
+//This actor is tasked with scheduling heartbeats.
+// To that end it sees every messages the system receives.
+// When a message is seen it schedule a heartbeat in the next heartbeatRate seconds.
+// Periodic heartbeats are sent with a period of messageDelay seconds.
+final case class Monitor(
+    heartbeatGenRef: ActorRef,
+    heartbeatRate: FiniteDuration,
+    messageDelay: FiniteDuration
+) extends Actor with ActorLogging with Timers {
+
+  // These keys are used to keep track of the timers states
+  private val periodicHbKey = 0
+  private val singleHbKey = 1
+
+  // Monitor is self-contained,
+  // To that end it doesn't know the ref of the connectionMediator
+  private var connectionMediatorRef = ActorRef.noSender
+
+  override def receive: Receive = LoggingReceive {
+
+    case Monitor.AtLeastOneServerConnected =>
+      connectionMediatorRef = sender()
+      if (!timers.isTimerActive(periodicHbKey))
+        timers.startTimerWithFixedDelay(periodicHbKey, TriggerHeartbeat, heartbeatRate)
+
+    case Monitor.NoServerConnected =>
+      timers.cancelAll()
+
+    case Monitor.TriggerHeartbeat =>
+      log.info("triggering a heartbeat")
+      timers.cancel(singleHbKey)
+      heartbeatGenRef ! Monitor.GenerateAndSendHeartbeat(connectionMediatorRef)
+
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      jsonRpcMessage.getParams match {
+
```

Comment:
Is this line return intended (even if it is a small pattern matching)?

Reply:
Fixed